_posts/2018-09-12-graphql-is-not-odata.md

```diff
@@ -0,0 +1,224 @@
+
+The GraphQL layer can then specifically optimize queries over the RESTful APIs, minimizing API calls necessary to resolve a query. Yes, it is work to implement each resolver, but the explicitness is so liberating. And your clients have no idea what is happening behind the GraphQL façade. One call from the client could fan out to a dozen APIs (in a good way*).  Orchestration stays out of the UI!
+
+* *Fanning out to a dozen APIs is good because it is still the server doing only what it was specifically built to do.*
+
+We land in a mode where the UI code is dumb. So dumb. It is just using boring fetch calls and dealing with boring JSON responses that already match the desired shape so no boilerplate deserialization is necessary. And then as the RESTful services change, get combined or divided, it is even more boring because the clients don’t even have to accommodate it. You only update the affected resolvers in GraphQL and the client is completely oblivious. Yawn.
+
+Another detail we have found important and highly successful: Our GraphQL layer IS NOT implemented or operated by the teams building RESTful services. The UI teams build that layer and Howard's team provides the platform and runs the service.
+
+This lets the service teams concentrate on REST and the domain model. GraphQL is an implementation detail of the UI layer—a technology chosen by UI, not services. This has avoided the whole REST vs. GraphQL debate with each of the dozens of service teams building APIs. They get to do their thing the way they want. For all they care, the UI consumes their services directly. We just happen to put a GraphQL server in between. We can centralize the GraphQL implementation into a smaller community of developers where we can foster reuse and commonalities more easily.
+
+Service teams are not burdened with *also* knowing and doing GraphQL. If they had to, they would inevitably look for ways to skip a layer, violating the separation of concerns. I often refer to using GraphQL as a having a double-paned window. UI cannot reach APIs directly; GraphQL is in between as insulation, reducing concerns on both sides. GraphQL helps both the UI and API layers as a result.
+
+### Downsides of GraphQL
+
+*[see in thread](https://twitter.com/JeffHandley/status/1035275129268236288)*
+
+Having to justify it when first adopting it or before enough service teams realize it will only benefit them.
+
+It is indeed yet another layer of ceremony where every single UI interaction needs to be thought through and implemented specifically (or covered by existing resolvers).
+
+It takes notable calendar time to design the schema thoughtfully such that it will have backwards compatibility as it grows over time. There is no versioning and the schema must be additive only (or with backwards compatible implementation detail changes). *Breaking changes are possible, but you must phase it: add new schema, rollout compensating changes to ALL clients, remove old schema.*
```

Comment:
I think this is the biggest stumbling block when starting a GraphQL effort - it forces you to think deeply about your data. Many organizations don't realize this going in and are frustrated when the effort doesn't work out or moves slowly.

Reply:
Added as a quote from you